awscli/botocore/utils.py

```diff
@@ -3305,6 +3339,231 @@ def fetch_token(
         )
 
 
+class SSOTokenFetcherAuth(BaseSSOTokenFetcher):
+    """Performs the authorization code grant with PKCE OAuth2.0 flow"""
+    _AUTH_GRANT_TYPES = ['authorization_code', 'refresh_token']
+    _AUTH_GRANT_DEFAULT_SCOPE = 'sso:account:access'
+
+    def __init__(
+            self, sso_region, client_creator, auth_code_fetcher, cache=None,
+            on_pending_authorization=None, time_fetcher=None
+    ):
+        super().__init__(sso_region, client_creator, cache,
+                         on_pending_authorization, time_fetcher
+                         )
+
+        self._auth_code_fetcher = auth_code_fetcher
+
+        # Generate the PKCE pair
+        self.code_verifier = ''.join(
+            random.SystemRandom().choice(
+                string.ascii_letters + string.digits + '-._~'
+            )
+            for _ in range(64)
```

Comment:
Left some notes about alternatives we might use for this verifier construction.

Reply:
Addressed internal discussion via https://github.com/aws/aws-cli/pull/8947/commits/e306e3ec67d3f8e3fb1d419f8d148f13f53f16a4